### PR TITLE
userspace: flesh out internal syscall interface

### PR DESCRIPTION
**Pull request by tyrantlizard77**
* Instead of a common system call entry function, we instead create a
table mapping system call ids to handler skeleton functions which are
invoked directly by the architecture code which receives the system
call.

* system call handler prototype specified. All but the most trivial
system calls will implement one of these. They validate all the
arguments, including verifying kernel/device object pointers, ensuring
that the calling thread has appropriate access to any memory buffers
passed in, and performing other parameter checks that the base system
call implementation does not check, or only checks with __ASSERT().

It's only possible to install a system call implementation directly
inside this table if the implementation has a return value and requires
no validation of any of its arguments.

A sample handler implementation for k_mutex_unlock() might look like:

```
u32_t _syscall_k_mutex_unlock(u32_t mutex_arg, u32_t arg2, u32_t arg3,
			      u32_t arg4, u32_t arg5, void *ssf)
{
	struct k_mutex *mutex = (struct k_mutex *)mutex_arg;
	_SYSCALL_ARG1;

	_SYSCALL_IS_OBJ(mutex, K_OBJ_MUTEX, 0,  ssf);
	_SYSCALL_ASSERT(mutex->lock_count > 0, ssf);
	_SYSCALL_ASSERT(mutex->owner == _current, ssf);

	k_mutex_unlock(mutex);

	return 0;
}
```

* the x86 port modified to work with the system call table instead of
calling a common handler function. fixed an issue where registers being
changed could confuse the compiler has been fixed; all registers, even
ones used for parameters, must be preserved across the system call.

* a new arch API for producing a kernel oops when validating system call
arguments added. The debug information reported will be from the system
call site and not inside the handler function.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

### CODE REVIEW

File: arch/x86/core/userspace.S [hunk 1]
```diff
@@ -8,46 +8,74 @@
...
+	push	%esp		/* syscall stack frame argument */
 	push	%edi		/* arg5 */
 	push	%ebx		/* arg4 */
-#ifndef CONFIG_X86_IAMCU
+
+	/* For IAMCU we don't need to push these three for the calling
+	 * convention but we need to make sure on return that these were the
+	 * same values when we got here as the compiler has no idea that the
+	 * 'int' instruction at the call site could have side effects.
+	 */
```
Comment (andrewboie) [inline arch/x86/core/userspace.S]: On retrospect I think this is wrong. I think I need to adjust the system call invocations to make some registers output operands

File: arch/x86/core/userspace.S [hunk 2]
```diff
@@ -8,46 +8,74 @@
...
- * further processing.
+ * further processing. We're on the kernel stack for the invoking thread.
  */
 SECTION_FUNC(TEXT, _x86_syscall_entry_stub)
-	push	%esi		/* call_id */
+	sti			/* re-enable interrupts */
+
+	/* Stash these values so we don't clobber them */
+	push	%ebp
+	push	%esi
```
Comment (andrewboie) [inline arch/x86/core/userspace.S]: can get rid of this if I set ESI to be an output operand in the syscall invoke macros

Comment (andrewboie) [reply arch/x86/core/userspace.S]: nope, nevermind, even if I specify ESI as an output operand the compiler still doesn't generate the right code. I think we have to leave like this, or at least re-visit later.

Comment (andrewboie) [reply arch/x86/core/userspace.S]: getting this right is tricky. will update PR once I think I have it.

File: arch/x86/core/fatal.c [hunk 3]
```diff
@@ -109,6 +109,27 @@ FUNC_NORETURN void _NanoFatalErrorHandler(unsigned int reason,
...
 }
 
+
+FUNC_NORETURN void _arch_syscall_oops(void *ssf_ptr)
+{
+	NANO_ESF oops_esf;
+	struct _x86_syscall_stack_frame *ssf;
+
+	ssf = (struct _x86_syscall_stack_frame *)ssf_ptr;
+	memset(&oops_esf, 0, sizeof(NANO_ESF));
```
Comment (lpereira) [inline arch/x86/core/fatal.c]: You can just use designated initializers here:

```C
struct _x86_syscall_stack_frame *ssf = ssf_ptr;
NANO_ESF oops_esf = {
    .eip = ssf->eip,
    .cs = ssf->cs,
    .eflags = ssf->eflags,
    .esp = ssf->cs == USER_CODE_SEG ? ssf->esp : 0,
};

_NanoFatalErrorHandler(_NANO_ERR_KERNEL_OOPS, &oops_esf);
```

Other aggregates will be initialized to 0, so you ditch the memset() call (which might even be removed by the compiler, but the code is cleaner anyway).

File: arch/x86/core/userspace.S [hunk 4]
```diff
@@ -8,46 +8,80 @@
...
 	push	%ecx		/* arg3 */
 	push	%edx		/* arg2	*/
 	push	%eax		/* arg1 */
 #endif
 
-	call	_k_syscall_entry
+	/* call_id is in ESI. bounds-check it, must be less than
+	 * K_SYSCALL_LIMIT
+	 */
+	cmp	$K_SYSCALL_LIMIT, %esi
```
Comment (lpereira) [inline arch/x86/core/userspace.S]: No need to do this now, but if the syscall table is limited to 256 elements, `_k_syscall_table` can be 256-elements long, with all pointers pointing to the oops syscall.  Then use just the lower 8 bits of `esi` to index the table. You'll save  a branch here.

File: arch/x86/core/userspace.S [hunk 5]
```diff
@@ -8,46 +8,80 @@
...
+	jae	_bad_syscall
+
+_id_ok:
+	/* from the call ID in ESI, load EBX with the actual function pointer
+	 * to call by looking it up in the system call dispatch table
+	 */
+	xor	%edi, %edi
+	mov	_k_syscall_table(%edi, %esi, 4), %ebx
```
Comment (lpereira) [inline arch/x86/core/userspace.S]: Much better than the proof of concept that had a switch statement.

File: arch/x86/core/userspace.S (hunk 4 above)
Comment (andrewboie) [reply arch/x86/core/userspace.S]: Might make this a Kconfig option. On X86 you need tons of RAM anyway due to page alignment but on an MPU-based system could be a toggle.

Comment (andyross) [reply arch/x86/core/userspace.S]: Counterargument: on Lakemont branches aren't nearly as expensive on bigcore (and this jump is forwards, so it's unlikely to be mispredicted in the non-error hot path) and the instruction to mask off the lower 8 bits (ESI has no byte subregisters) is actually going to be longer than the jump, polluting the instruction cache.  Not sure how I'd put money down on which was worse.

File: arch/x86/core/userspace.S [hunk 6]
```diff
@@ -8,46 +8,80 @@
...
+
+	/* Run the handler, which is some entry in _k_syscall_table */
+	call	*%ebx
 
 	/* EAX now contains return value. Pop or xor everything else to prevent
 	 * information leak from kernel mode.
 	 */
 #ifndef CONFIG_X86_IAMCU
-	pop	%edx	/* old EAX value, discard it */
+	/* SYS V convention */
```
Comment (andyross) [inline arch/x86/core/userspace.S]: Same whine about SysV == !IAMCU being obvious.

File: arch/x86/core/userspace.S [hunk 7]
```diff
@@ -8,46 +8,80 @@
...
+#endif
 	iret
 
+_bad_syscall:
+	/* ESI had a bogus syscall value in it, replace with the bad syscall
+	 * handler's ID.  This clobbers ESI but the bad syscall handler never
+	 * returns anyway, it's going to generate a kernel oops
+	 */
+	mov	$K_SYSCALL_BAD, %esi
+	jmp	_id_ok
```
Comment (andyross) [inline arch/x86/core/userspace.S]: Clever and small.  I like it.

File: arch/x86/core/fatal.c [hunk 8]
```diff
@@ -109,6 +109,25 @@ FUNC_NORETURN void _NanoFatalErrorHandler(unsigned int reason,
...
+	struct _x86_syscall_stack_frame *ssf =
+		(struct _x86_syscall_stack_frame *)ssf_ptr;
+	NANO_ESF oops_esf = {
+		.eip = ssf->eip,
+		.cs = ssf->cs,
+		.eflags = ssf->eflags
+	};
+
+	if (oops_esf.cs == USER_CODE_SEG) {
+		oops_esf.esp = ssf->esp;
```
Comment (andyross) [inline arch/x86/core/fatal.c]: Doesn't this need to set the stack segment field too?

File: include/syscall.h [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+typedef u32_t (*_k_syscall_handler_t)(u32_t arg1, u32_t arg2, u32_t arg3,
+				      u32_t arg4, u32_t arg5, void *ssf);
+
+
+extern const _k_syscall_handler_t _k_syscall_table[K_SYSCALL_LIMIT];
+
+#define _SYSCALL_ASSERT(expr, ssf) \
+	do { \
+		if (!(expr)) { \
+			printk("system call failed: " #expr "\n"); \
```
Comment (andyross) [inline include/syscall.h]: Clarity: "system call failed" implies that this is an application logging a routine runtime failure when it's actually a fatal error.

Naming: calling this "ASSERT" implies to the reader that it's debug code that will be removed in an optimized build and potentially removable when maintaining code, when it's actually a critical piece of the security model.  Suggest "SYSCALL_VALIDATE" or somesuch.

**Review by andyross — CHANGES_REQUESTED** [review]
Mostly whining.  One spot where I think you're missing a field in the error stack frame.

File: arch/x86/core/fatal.c (hunk 8 above)
Comment (andrewboie) [reply arch/x86/core/fatal.c]: NANO_ESF has no SS field.

File: arch/x86/core/fatal.c [hunk 10]
```diff
@@ -109,6 +109,25 @@ FUNC_NORETURN void _NanoFatalErrorHandler(unsigned int reason,
...
+	}
+
+	_NanoFatalErrorHandler(_NANO_ERR_KERNEL_OOPS, &oops_esf);
+}
+
+
```
[the remark this answers is not in the document]
Comment (andrewboie) [reply arch/x86/core/fatal.c]: for some reason I always thought the standard was to have two blank lines in between functions, but I looked at the Linux coding style document and indeed, its just one line. will fix

File: include/syscall.h (hunk 9 above)
Comment (andrewboie) [reply include/syscall.h]: sure I'll rename it

File: arch/x86/core/fatal.c (hunk 8 above)
Comment (andrewboie) [reply arch/x86/core/fatal.c]: Also note, we have plans to cleanup and centralize _NanoFatalErrorHandler(). Currently it dumps out everything in NANO_ESF, even if some fields are unknown (it prints 0 or 0xdeadadead in these cases which is not ideal). in situations like this we ought to skip the register dump.

**Review by andyross — DISMISSED** [review]

**Comment by andrewboie** [comment]
Updated PR:
1) _SYSCALL_ASSERT renamed to _SYSCALL_VERIFY
2) extraneous comments / whitespace issues fixed
3) bad system calls now report the offending ID passed in
4) improved debug message when _SYSCALL_VERIFY fails

File: include/syscall.h [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+ *
+ * SPDX-License-Identifier: Apache 2.0
+ */
+
+
+#ifndef _ZEPHYR_SYSCALL_H_
+#define _ZEPHYR_SYSCALL_H_
+
+/* Fixed system call IDs. We use #defines instead of enumeration so that if
+ * system calls are retired it does not shift the IDs of other system calls.
```
Comment (lpereira) [inline include/syscall.h]: Do we care if syscall IDs are recycled? If someone rebuilds an application with the retired syscall, it should use the new numbers anyway.

Comment (andrewboie) [reply include/syscall.h]: There will come a time where the application and kernel can be built and loaded separately, we already have requests for it, and I don't want to impede that. My feeling is that this type of interaction between the app and kernel should be graven in stone.

**Review by andyross — APPROVED** [review]

**Pull request merged**